lib/galaxy/model/migrate/versions/0155_add_content_update_time.py

```diff
@@ -0,0 +1,105 @@
+"""
+Adds trigger to dataset table to update history.update_time when
+contents are changed.
+"""
+from __future__ import print_function
+
+import logging
+
+from sqlalchemy import DDL, MetaData
+
+log = logging.getLogger(__name__)
+metadata = MetaData()
+
+
+def upgrade(migrate_engine):
+    print(__doc__)
+    metadata.bind = migrate_engine
+    metadata.reflect()
+    drop_update_trigger(migrate_engine)
+    install_update_trigger(migrate_engine)
+
+
+def downgrade(migrate_engine):
+    metadata.bind = migrate_engine
+    metadata.reflect()
+    drop_update_trigger(migrate_engine)
+
+
+def not_pg(self, target, bind, state, **kw):
+    return bind.engine.name != 'postgresql'
+
+
+def install_update_trigger(migrate_engine):
+    """Installs trigger on database table to update history table
+    when contents have changed. Installs a function and a trigger
+    for postgres, other sql variants only require the trigger def
+    """
+
+    pg_create_trigger = DDL("""
+        CREATE FUNCTION update_history_content_update_time()
+            RETURNS trigger
+            LANGUAGE 'plpgsql'
+        AS $BODY$
+        begin
+            update history h
+            set update_time = current_timestamp
+            from history_dataset_association hda
+            where h.id = hda.history_id
+            and hda.dataset_id = NEW.id;
+            return NEW;
+        end;
+        $BODY$;
+        CREATE TRIGGER update_history_update_time
+            BEFORE INSERT OR DELETE OR UPDATE
+            ON dataset
+            FOR EACH ROW
+            EXECUTE PROCEDURE update_history_content_update_time();
+    """).execute_if(dialect='postgresql')
+
+    pg_create_trigger.execute(bind=migrate_engine)
+
+    # Looks like sqlite doesn't like multiple actions in some
+    # variants, so we build 3 triggers
+    build_trigger('INSERT').execute(bind=migrate_engine)
+    build_trigger('UPDATE').execute(bind=migrate_engine)
+    build_trigger('DELETE').execute(bind=migrate_engine)
+
+
+def drop_update_trigger(migrate_engine):
+    """Drops trigger on dataset table."""
+
+    pg_drop_trigger = DDL("""
+        DROP TRIGGER IF EXISTS update_history_update_time ON dataset;
+        DROP FUNCTION IF EXISTS update_history_content_update_time();
+    """).execute_if(dialect='postgresql')
+
+    pg_drop_trigger.execute(bind=migrate_engine)
+    build_drop_trigger("INSERT").execute(bind=migrate_engine)
+    build_drop_trigger("UPDATE").execute(bind=migrate_engine)
+    build_drop_trigger("DELETE").execute(bind=migrate_engine)
+
+
+def build_trigger(op):
+    create_trigger_template = """
+        CREATE TRIGGER BEFORE_{operation}_DATASET
+            BEFORE {operation} ON dataset
```

Comment:
Why not
```suggestion
            AFTER {operation} ON dataset
```
? It makes more sense to me to update the history timestamp after the change to its content.

Reply:
Sure, I guess an error could get thrown or something. After might be more appropriate.

Reply:
@Nerdinacan @nsoranzo aren't all these supposed to be `AFTER`? Current code uses `BEFORE`. Unless there's been more discussion and it's been agreed to use `BEFORE`, in which case please ignore this comment.

Reply:
@ic4f It was change back to `BEFORE` by @Nerdinacan in commit 335499fc5449dfd7ff38104ac84f5e522220425c , not sure why.

Reply:
Oh, OK, thanks for checking @nsoranzo